pkg/graveler/graveler.go

```diff
@@ -1597,45 +1597,64 @@ func CommitExists(ctx context.Context, repositoryID RepositoryID, commitID Commi
 }
 
 func (g *KVGraveler) AddCommitToBranchHead(ctx context.Context, repositoryID RepositoryID, branchID BranchID, commit Commit) (CommitID, error) {
-	// TODO(issue 3569) - kv implementation optimistic lock commit
-	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
-		// parentCommitID should always match the HEAD of the branch.
-		// Empty parentCommitID matches first commit of the branch.
-		parentCommitID, err := validateCommitParent(ctx, repositoryID, commit, g.RefManager)
-		if err != nil {
-			return nil, err
-		}
+	repo, err := g.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", err
+	}
 
-		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
-		if err != nil {
-			return nil, err
-		}
+	// parentCommitID should always match the HEAD of the branch.
+	// Empty parentCommitID matches first commit of the branch.
+	parentCommitID, err := validateCommitParent(ctx, repositoryID, commit, g.RefManager)
+	if err != nil {
+		return "", err
+	}
+
+	// verify access to meta range
+	ok, err := g.CommittedManager.Exists(ctx, repo.StorageNamespace, commit.MetaRangeID)
+	if err != nil {
+		return "", fmt.Errorf("commit missing meta range %s: %w", commit.MetaRangeID, err)
+	}
+	if !ok {
+		return "", fmt.Errorf("%w: %s", ErrMetaRangeNotFound, commit.MetaRangeID)
+	}
+
+	// add commit and/or use the commit id of the one we found
+	commitID, err := g.RefManager.AddCommit(ctx, repositoryID, commit)
+	if err != nil && !errors.Is(err, kv.ErrPredicateFailed) {
+		return "", fmt.Errorf("adding commit: %w", err)
+	}
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
 		if branch.CommitID != parentCommitID {
 			return nil, ErrCommitNotHeadBranch
 		}
 
-		// check if commit already exists.
-		commitID := CommitID(ident.NewHexAddressProvider().ContentAddress(commit))
-		if exists, err := CommitExists(ctx, repositoryID, commitID, g.RefManager); err != nil {
-			return nil, err
-		} else if exists {
-			return commitID, nil
+		// report conflict if there are changes pending on the branch
+		if len(branch.SealedTokens) > 0 {
```

Comment:
Not necessarily changes - since the sum of the staging area might be empty. Maybe use stagingEmpty instead?